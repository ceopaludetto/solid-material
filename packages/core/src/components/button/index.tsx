import type { VariantProps } from "class-variance-authority";
import type { JSX } from "solid-js";

import { Button as KButton } from "@kobalte/core";
import { Show, splitProps } from "solid-js";

import { icon, root } from "./styles";
import { createRipples } from "~/primitives";
import { mergeWithRefs } from "~/utils/refs";

export type ButtonProps = KButton.ButtonRootOptions &
  Omit<VariantProps<typeof root>, "hasEndIcon" | "hasStartIcon"> & {
    class?: string;
    children: JSX.Element;
    startIcon?: JSX.Element;
    endIcon?: JSX.Element;
  };

export function Button(props: ButtonProps) {
  const [trigger] = createRipples({ disabled: props.disabled });
  const [local, rest] = splitProps(props, ["class", "children", "variant", "startIcon", "endIcon"]);

  return (
    <KButton.Root
      class={root({
        class: local.class,
        variant: local.variant ?? "filled",
        hasStartIcon: !!local.startIcon,
        hasEndIcon: !!local.endIcon,
      })}
      {...mergeWithRefs(trigger, rest)}
    >
      <Show when={local.startIcon}>{(startIcon) => <span class={icon()}>{startIcon()}</span>}</Show>
      {local.children}
      <Show when={local.endIcon}>{(endIcon) => <span class={icon()}>{endIcon()}</span>}</Show>
    </KButton.Root>
  );
}
