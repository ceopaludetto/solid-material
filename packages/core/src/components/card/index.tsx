import type { As, PolymorphicProps } from "@kobalte/utils";
import type { VariantProps } from "class-variance-authority";
import type { JSX } from "solid-js";

import { splitProps } from "solid-js";
import { Dynamic } from "solid-js/web";

import { root } from "./styles";

type CardOwnProps = VariantProps<typeof root> & {
  class?: string;
  children?: JSX.Element;
};

export type CardProps<T extends As = "div"> = PolymorphicProps<T, CardOwnProps>;

export function Card<T extends As = "div">(props: CardProps<T>) {
  const [local, rest] = splitProps(props, ["as", "children", "class", "variant"]);

  return (
    <Dynamic
      class={root({ class: local.class, variant: local.variant ?? "filled" })}
      component={local.as ?? "div"}
      {...rest}
    >
      {local.children}
    </Dynamic>
  );
}
