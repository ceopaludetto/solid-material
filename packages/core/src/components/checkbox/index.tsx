import type { As, PolymorphicProps } from "@kobalte/utils";
import type { VariantProps } from "class-variance-authority";
import type { JSX } from "solid-js";

import { Checkbox as KCheckbox } from "@kobalte/core";
import { createSignal, Show, splitProps } from "solid-js";

import { control, indicator, root } from "./styles";
import { createRipples } from "~/primitives";
import { mergeWithRefs } from "~/utils/refs";

type CheckboxOwnProps = KCheckbox.CheckboxRootOptions &
  VariantProps<typeof root> & {
    label: string;
    checkIcon: JSX.Element;
    indeterminateIcon: JSX.Element;
    class?: string;
  };

export type CheckboxProps<T extends As = "label"> = PolymorphicProps<T, CheckboxOwnProps>;

export function Checkbox<T extends As = "label">(props: CheckboxProps<T>) {
  const [ref, getRef] = createSignal<HTMLLabelElement>();

  let controlRef!: HTMLDivElement;

  createRipples({ ref, positioner: () => controlRef, size: 40, center: true, disabled: props.isDisabled });
  const [local, rest] = splitProps(props, ["class", "label", "checkIcon", "indeterminateIcon", "labelPlacement"]);

  return (
    <KCheckbox.Root
      class={root({ class: local.class, labelPlacement: local.labelPlacement })}
      {...mergeWithRefs(getRef, rest)}
    >
      <KCheckbox.Input />
      <KCheckbox.Label>{local.label}</KCheckbox.Label>
      <KCheckbox.Control ref={controlRef} class={control()}>
        <KCheckbox.Indicator forceMount class={indicator()}>
          <Show when={!props.isIndeterminate}>{local.checkIcon}</Show>
          <Show when={props.isIndeterminate}>{local.indeterminateIcon}</Show>
        </KCheckbox.Indicator>
      </KCheckbox.Control>
    </KCheckbox.Root>
  );
}
