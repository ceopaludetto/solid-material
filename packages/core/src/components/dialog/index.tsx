import type { As, PolymorphicProps } from "@kobalte/utils";
import type { JSX } from "solid-js";

import { Dialog as KDialog } from "@kobalte/core";
import { Show, splitProps } from "solid-js";

import { actions, content, description, icon, overlay, positioner, title } from "./styles";

type DialogOwnProps = KDialog.DialogRootOptions & {
  class?: string;
  children?: JSX.Element;
  title: string;
  description: string;
  endAdornment?: JSX.Element;
  icon?: JSX.Element;
};

export type DialogContentProps<T extends As = "div"> = PolymorphicProps<T, DialogOwnProps>;

function Content<T extends As = "div">(props: DialogContentProps<T>) {
  const [local, rest] = splitProps(props, ["class", "title", "children", "description", "endAdornment", "icon"]);

  return (
    <KDialog.Root {...rest}>
      <Show when={local.children}>{local.children}</Show>
      <KDialog.Portal>
        <KDialog.Overlay class={overlay()} />
        <div class={positioner()}>
          <KDialog.Content class={content({ class: local.class })}>
            <Show when={local.icon}>
              <span class={icon()}>{local.icon}</span>
            </Show>
            <KDialog.Title class={title()}>{local.title}</KDialog.Title>
            <KDialog.Description class={description()}>{local.description}</KDialog.Description>
            <div class={actions()}>
              <Show when={!!local.endAdornment}>{local.endAdornment}</Show>
            </div>
          </KDialog.Content>
        </div>
      </KDialog.Portal>
    </KDialog.Root>
  );
}

export const Dialog = { Content, Trigger: KDialog.Trigger };
