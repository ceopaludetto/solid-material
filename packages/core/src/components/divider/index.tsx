import { Separator as KSeparator } from "@kobalte/core";
import { splitProps } from "solid-js";

import { root } from "./styles";

export type DividerProps = KSeparator.SeparatorRootOptions & {
  class?: string;
};

export function Divider(props: DividerProps) {
  const [local, rest] = splitProps(props, ["class"]);

  return <KSeparator.Root class={root({ class: local.class })} {...rest} />;
}
