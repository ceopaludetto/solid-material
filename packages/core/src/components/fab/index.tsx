import type { PolymorphicProps, As } from "@kobalte/utils";
import type { VariantProps } from "class-variance-authority";
import type { JSX } from "solid-js";

import { Button as KButton } from "@kobalte/core";
import { createSignal, createMemo, Show, splitProps } from "solid-js";

import { icon, root } from "./styles";
import { createRipples } from "~/primitives";
import { mergeWithRefs } from "~/utils/refs";

type FABOwnProps = KButton.ButtonRootOptions &
  VariantProps<typeof root> & {
    class?: string;
    icon?: JSX.Element;
    children?: JSX.Element;
  };

export type FABProps<T extends As = "button"> = PolymorphicProps<T, FABOwnProps>;

export function FAB<T extends As = "button">(props: FABProps<T>): JSX.Element {
  const [ref, getRef] = createSignal<HTMLButtonElement>();

  createRipples({ ref, disabled: props.isDisabled });
  const [local, rest] = splitProps(props, ["class", "children", "variant", "icon", "size"]);

  const isExtended = createMemo(() => !!local.children);

  return (
    <KButton.Root
      class={root({
        class: local.class,
        variant: local.variant ?? "surface",
        size: isExtended() ? "medium" : local.size ?? "medium",
        isExtended: isExtended(),
      })}
      {...mergeWithRefs(getRef, rest)}
    >
      <Show when={!!local.icon}>
        <span class={icon({ size: isExtended() ? "medium" : local.size ?? "medium" })}>{local.icon}</span>
      </Show>
      {local.children}
    </KButton.Root>
  );
}
