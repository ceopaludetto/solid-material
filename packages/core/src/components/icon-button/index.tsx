import type { VariantProps } from "class-variance-authority";
import type { JSX } from "solid-js";

import { Button as KButton } from "@kobalte/core";
import { splitProps } from "solid-js";

import { root } from "./styles";
import { createRipples } from "~/primitives";
import { mergeWithRefs } from "~/utils/refs";

export type IconButtonProps = KButton.ButtonRootOptions &
  VariantProps<typeof root> & {
    class?: string;
    children: JSX.Element;
  };

export function IconButton(props: IconButtonProps) {
  const [trigger] = createRipples({ center: true, disabled: props.disabled });
  const [local, rest] = splitProps(props, ["children", "class", "variant"]);

  return (
    <KButton.Root
      class={root({ class: local.class, variant: local.variant ?? "filled" })}
      {...mergeWithRefs(trigger, rest)}
    >
      {local.children}
    </KButton.Root>
  );
}
