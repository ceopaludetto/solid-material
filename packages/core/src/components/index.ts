export * from "./button";
export * from "./card";
export * from "./checkbox";
export * from "./dialog";
export * from "./divider";
export * from "./fab";
export * from "./icon-button";
export * from "./navigation-drawer";
export * from "./navigation-drawer-item";
export * from "./navigation-item";
export * from "./navigation-rail";
export * from "./switch";
export * from "./tabs";
export * from "./text-field";
export * from "./top-bar";
