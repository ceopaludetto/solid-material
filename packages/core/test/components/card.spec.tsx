import { render } from "@solidjs/testing-library";
import { describe, expect, it } from "vitest";

import { Card } from "~/components";

describe("Card", () => {
  it("should render", () => {
    const { asFragment } = render(() => <Card>Test</Card>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with outlined variant", () => {
    const { asFragment } = render(() => <Card variant="outlined">Test</Card>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render as custom element", () => {
    const { asFragment, getByText } = render(() => <Card as="button">Test</Card>);

    expect(asFragment()).toMatchSnapshot();
    expect(getByText(/Test/i)).toBeInstanceOf(HTMLButtonElement);
  });
});
