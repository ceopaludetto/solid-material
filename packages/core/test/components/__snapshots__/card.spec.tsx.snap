// Vitest Snapshot v1

exports[`Card > should render 1`] = `"<div class=\\"rounded-md bg-surface-variant\\">Test</div>"`;

exports[`Card > should render as custom element 1`] = `"<button class=\\"rounded-md bg-surface-variant\\">Test</button>"`;

exports[`Card > should render with outlined variant 1`] = `"<div class=\\"rounded-md border border-outline bg-surface\\">Test</div>"`;
